### ISSUE TITLE
How is it possible to discharge Show instance of Sigma?

### ISSUE THREAD

**Issue by tomjaguarpaw** [goldfirere/singletons#579]
Consider the program below.  I think it's uncontroversial (although there's a possibility that I've misunderstood how `Apply` instances are supposed to be written). This is probably the simplest possible program of this sort, yet I don't see how I can `show mySigma`.

The error message mentions `Show (Apply Id x)` but what it really wants is `forall (x :: ()). Show (Apply Id x)` (via `ShowApply`/`ShowApply'`). I don't see how that can possibly be discharged.

Is there an example somewhere showing how this `Show` instance is supposed to be used in practice?

```haskell
data SUnit (a :: ()) where
  MkSUnit :: SUnit '()

type instance Sing = SUnit

instance SingI '() where
  sing = MkSUnit

instance Show (SUnit a) where
  show MkSUnit = "MkSUnit"

type Id :: () ~> Type
data Id t = MkId

type instance Apply Id '() = ()

mySigma :: Sigma () Id
mySigma = MkSUnit :&: ()

example = show mySigma
```

----

```
test18.hs:111:11: error:
    • No instance for (Show (Apply Id x)) arising from a use of ‘show’
    • In the expression: show mySigma
      In an equation for ‘example’: example = show mySigma
    |
111 | example = show mySigma
    |           ^^^^^^^^^^^^
```

**Comment by RyanGlScott** [comment]
Yeah, this is admittedly somewhat unsatisfactory. I adapted the implementation of `Sigma`'s `Show` instance from a trick that I use to define `Show` instances for singleton types, which is described at length [here](https://hackage.haskell.org/package/singletons-3.0.2/docs/Data-Singletons-ShowSing.html). In both types of `Show` instances, I have to discharge constraints that look like this:

```hs
forall a. Show (f a)
```

In the case of singleton types, `f` is something like `SUnit`, and it is quite easy to discharge a `forall a. Show (SUnit a)` constraint, since we can define a `Show (SUnit a)` instance with no other constraints on the `a`.

In the case of sigma types, however, `f` is something `Apply Id`. It's trickier to resolve a `forall a. Show (Apply Id a)` constraint, as we must be able to reduce `Apply Id a` to something `Show`-able _without knowing what `a` is_. That is not possible with the definition of `Id` that you give:

```hs
type instance Apply Id '() = ()
```

But note that if you change this code such that it no longer scrutinizes the second argument:

```
type instance Apply Id _ = ()
```

Then `Apply Id a` evaluates to `()` for all `a`, and then your code typechecks:

```
λ> example
"MkSUnit :&: ()"
```

But yes, this is rather limited in practice, since most type families will actually want to scrutinize the argument in some way, thereby defeating this trick. Sadly, I don't know of another way to define this `Show` instance.

**Comment by RyanGlScott** [comment]
That's quite clever. A bit of shame that we'd need to derive an instance of yet another type class for everything, but perhaps that can't be helped. (Unless there's a way to define `ConstraintFromSing` in terms of existing `singletons` classes, but I can't figure out a way to do so.)

**Comment by tomjaguarpaw** [comment]
Ah, it seems that it can work even if `Apply` scrutinises its argument, as long as it returns something uniform that can be universally shown, such as a GADT:

```haskell
data Foo (a :: ()) where
  MkFoo :: Foo '()

type FooT :: () ~> Type
data FooT t = MkFooT

instance Show (Foo a) where
  show MkFoo = "MkFoo"

type instance Apply FooT t = Foo t

mySigma :: Sigma () FooT
mySigma = MkSUnit :&: MkFoo

example = show mySigma
```

**Comment by tomjaguarpaw** [comment]
I see, fair enough, in which case the key is to return something with a universal `Show` instance. It turns out that's still too restrictive in practice, as generally you want to use it on something with a universal show instance constrained by `SingI a =>`.

**Comment by RyanGlScott** [comment]
> Instead of `ConstraintFromSing Show s t` could we have `forall a. SingI a => Show (t @@ i)`? We do something similar in our work codebase.

Oh, that is an interesting idea. Can you write out a full `Show` instance for `Sigma` to demonstrate how this works? (I'm sure it's possible, but it's a bit tricky to use a quantified constraint with a type family such as `@@` in the head of the constraint.)

> It works mostly fine, except it would be useful if we had https://gitlab.haskell.org/ghc/ghc/-/issues/14822.

I'm not sure what you mean here. Can you elaborate a bit more?

**Comment by RyanGlScott** [comment]
Ah, I see what you mean now. Indeed, if you used a `forall (a :: t). SingI a => CApply Show f a` constraint in the `Show` instance for `Sigma`, then you still wouldn't be able to call `show mySigma` on your original `mySigma :: Sigma () Id` example from https://github.com/goldfirere/singletons/issues/579#issue-2019096984, since GHC would still complain thusly:

```
    • Could not deduce (Show (Apply Id a)) arising from a use of ‘show’
      from the context: SingI a
```

Since the quantified constraint requires `SingI (a :: ())`, it feels like one ought to be able to match on `sing :: Sing '()` and convince GHC that `a ~ '()`. This would cause `Apply Id '()` to reduce to `()` and all would be well. But as you say, it's not straightforward to do this at the level of constraints, which is where GHC#14822 enters the picture. I think I'm following now.
